### PR TITLE
Done DLPAR rework for dedicated and shared mode.

### PR DESCRIPTION
**Pull request by SamirMulani**
Below code cahages are done,
-> Added support for avocado ssh library.
-> Added support to configure the DLPAR operation using .yaml file and generate tests.cfg file. -> The pexpect library's method of executing commands has been removed. -> Done code cleanup.
-> Added new README.md

Signed-off-by: Samir <samir@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by abdhaleegit** [comment]
@SamirMulani Could you please paste here the overall flow , may be block diagram of dlpar flow and also what different dlpar it supports ?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
Logs please .. , also its a huge change, please split this into mulitple commits

**Comment by SamirMulani** [comment]
> @SamirMulani Could you please paste here the overall flow , may be block diagram of dlpar flow and also what different dlpar it supports ?

**Dlpar Implementation details:**

1. Added support for avocado ssh library.
-> To execute the HMC commands in the old implementation, we used pexpect's python library for ssh operations. In the new implementation, we are using avocado's ssh library instead.
-> The implementation code for the pexpect way of communicatinon has been removed from dlpar_api/api.py.
**Changed files:**
1.1 dlpar/dlpar_api/api.py
1.2 dlpar/cpu_unit.py
1.3 dlpar/dedicated_cpu.py
1.4 dlpar/mem.py

2. Added support to configure the DLPAR operation using .yaml file and generate tests.cfg file.
-> The old implementation required maintaining a single test.cfg file for dedicated and shared mode operation. Hence to avoid maintaining one extra config file, we removed the test.cfg file from the config folder. Added support to configure the dlpar operation via a .yaml file and to generate the test.cfg file dynamically based on that .yaml file.
**Changed files:**
2.1 dlpar/dlpar_main.py

 
3. Added new README.md.
**Changed files:**
3.1 dlpar/README.md

**Dlpar-Operation supports:**
1. Dedicated mode cpu
-> ADD
-> REMOVE
-> MOVE

2. Dedicated mode mem
-> ADD
-> REMOVE
-> MOVE

3. Shared mode cpu
-> ADD
-> REMOVE
-> MOVE

4. Shared mode mem
-> ADD
-> REMOVE
-> MOVE


DLPAR

**Comment by SamirMulani** [comment]
> Logs please .. , also its a huge change, please split this into mulitple commits

```
[root]# avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : e3cad3d51de63503f71d6c79a0ddd1fbf4492fe7
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-01-19T23.52-e3cad3d/job.log
 (1/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-cpu_dlpar-d527: STARTED
 (1/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-cpu_dlpar-d527: PASS (321.45 s)
 (2/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-mem_dlpar-355b: STARTED
 (2/2) dlpar_main.py:DlparTests.test_dlpar;run-config-lpar_mode-dedicated-test_type-mem_dlpar-355b: PASS (64.11 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-01-19T23.52-e3cad3d/results.html
JOB TIME   : 387.95 s
```

File: dlpar/dlpar_main.py [hunk 1]
```diff
@@ -55,105 +62,78 @@ def run_cmd(self, test_cmd, dlpar_type_flag=""):
...
+                Sha_obj.remove_proc()
+                Sha_obj.move_proc()
+
+    def test_mem_dlpar(self):
+        Mem_obj = Memory(self.sorted_payload, log='memory.log')
+        for i in range(self.iterations):
+            Mem_obj.mem_add()
+            Mem_obj.mem_rem()
+            Mem_obj.mem_move()
+            # Mem_obj.mem_mix_ope()
```
Comment (abdhaleegit) [inline dlpar/dlpar_main.py:139]: so the same test will also cover shared scenarios ?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
@SamirMulani Thank you for rework, the code looks more readable and structured now.. all looks good except few minor comments

File: dlpar/dlpar_main.py (hunk 1 above)
Comment (PraveenPenguin) [reply dlpar/dlpar_main.py:139]: mem does not have shared concept

File: dlpar/README.md [hunk 2]
```diff
@@ -13,68 +13,31 @@ OR
...
-#end - config/test.cfg
+# Sample dlpar.yaml is provided as below:
+#cat dlpar_main.py.data/dlpar.yaml
+
+cfg_cpu_per_proc: 8
+hmc_partition:   <partition name, for hmc it is same as managed system name>
+hmc_user:        <hmc default user, ex: hscroot>
+hmc_passwd:      <hscroot password>
+target_lpar_name: <fully qualified  lpar name>
+target_partition: <partition name, should be same as lpar name>
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply dlpar/README.md:26]: rename to lpar_hostname

File: dlpar/dlpar_main.py.data/dlpar.yaml [hunk 3]
```diff
@@ -1,20 +1,20 @@
+cfg_cpu_per_proc: 8
+hmc_partition:
```
Comment (abdhaleegit) [inline dlpar/dlpar_main.py.data/dlpar.yaml:2]: hmc_manageSystem

File: dlpar/dlpar_main.py.data/dlpar.yaml [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
-            mode: 'sha_smt_workload'
-        cpu_fold_sha:
-            mode: 'sha_cpu_fold_workload'
-    test_type: !mux
-         cpu_dlpar:
-             test_case: 'cpu'
-         mem_dlpar:
-             test_case: 'mem'
-         
+            lp_mode: 'dedicated'
```
Comment (abdhaleegit) [inline dlpar/dlpar_main.py.data/dlpar.yaml:20]: separate shared and dedicated

**Review by abdhaleegit — CHANGES_REQUESTED** [review]

File: dlpar/dlpar_api/api.py [hunk 5]
```diff
@@ -50,90 +47,47 @@ def __str__(self):
...
+            self.user = config_payload.get('target_user')
+            self.passwd = config_payload.get('target_passwd')
+            self.partition = config_payload.get('target_partition')
+            self.machine = config_payload.get('hmc_partition')
+        else:
+            self.name = config_payload.get('hmc_name')
+            self.user = config_payload.get('hmc_user')
+            self.passwd = config_payload.get('hmc_passwd')
+            self.partition = config_payload.get('hmc_partition')
+            self.machine = config_payload.get('hmc_partition')
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:74]: machine and partition are assigned same value. This seems incorrect. Can you check?

File: dlpar/dlpar_api/api.py [hunk 6]
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
...
+                ' ' + dlpar_flag[2] + ' ' + str(quantity) + \
+                ' -p "' + linux_machine.partition + '"' + ' -w 0 '
+        elif dlpar_flag[1] == "m":
+            cmd = 'chhwres -m ' + linux_machine[0].machine + \
+                ' -r ' + dlpar_flag[0] + ' -o ' + dlpar_flag[1] + ' ' + \
+                dlpar_flag[2] + ' ' + str(quantity) + \
+                ' -p "' + linux_machine[0].partition + '"' + \
+                ' -t "' + linux_machine[1].partition + '"' + ' -w 0 '
+        else:
+            self.log.error("Envalid DLPAR flag")
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:213]: Typo, Envalid --> Invalid

File: dlpar/dlpar_api/api.py [hunk 7]
```diff
@@ -254,20 +418,37 @@ def get_cpu_option(self, linux_machine, option):
...
         return opt_value
 
+    def check_memory_proc(self, machine):
+        """
+        Check on a linux client how much physical memory we have.
+        """
+        memory = 0
+        p_cmd = "cat /proc/meminfo | grep MemTotal | awk -F' ' '{ print $2 }'"
+        self.log.info('Checking memory under linux on %s' % machine.partition)
+        self.log.debug('Checking memory under linux on %s' % machine.partition)
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:434]: One of the logging statement can be removed.

File: dlpar/dlpar_api/api.py [hunk 8]
```diff
@@ -254,20 +418,37 @@ def get_cpu_option(self, linux_machine, option):
...
             linux_machine.partition
         self.log.debug(d_msg)
         return opt_value
 
+    def check_memory_proc(self, machine):
+        """
+        Check on a linux client how much physical memory we have.
+        """
+        memory = 0
+        p_cmd = "cat /proc/meminfo | grep MemTotal | awk -F' ' '{ print $2 }'"
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:432]: What is the need for this statement? I see the p_cmd value is overwritten in the try block below. Only one statement should suffice.

File: dlpar/dlpar_api/api.py [hunk 9]
```diff
@@ -304,7 +485,9 @@ def linux_check_rm_cpu(self, linux_machine, quantity_before,
 
         # Also, verify if /proc/cpuinfo show all processors
         c_cmd = 'cat /proc/cpuinfo'
-        cmd = linux_machine.sshcnx.run_command(c_cmd).decode()
+        cmd = process.run(c_cmd).stdout()
+        self.log.debug(cmd)
+        # cmd = linux_machine.sshcnx.cmd(c_cmd).decode()
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:490]: any reason to keep this statement? If not can you remove it?

File: dlpar/dlpar_api/api.py [hunk 10]
```diff
@@ -451,3 +629,552 @@ def set_virtual_proc_and_proc_units(self, linux_machine, ideal_procs,
...
+
+    def __init__(self, config_payload, log='dedicated_cpu.log'):
+
+        TestCase.__init__(self, log, 'Dedicated CPU', config_payload)
+
+        # Get test configuration
+        self.quant_to_test = 2
+        self.quant_to_test = config_payload.get('ded_quantity_to_test')
+        self.sleep_time = 60
+        self.sleep_time = config_payload.get('sleep_time')
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:656]: These variables are assigned values in succession. The first assignment is unnecessary.

File: dlpar/dlpar_api/api.py [hunk 11]
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
...
+            curr_procs_before = curr_procs_info
+            a_msg = 'Adding %d dedicated cpus to partition %s.' % \
+                    (quantity, linux_machine.partition)
+            a_condition = int(self.get_cpu_option(
+                linux_machine, 'curr_procs')) == \
+                curr_procs_before + quantity
+            if not self.log.check_log(a_msg, a_condition, False):
+                e_msg = 'Error adding %d dedicated cpus to partition %s.' % \
+                        (quantity, linux_machine.partition)
+
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:234]: Remove the blank line.

File: dlpar/dlpar_api/api.py [hunk 12]
```diff
@@ -204,46 +154,260 @@ def check_log(self, message, code, die=True):
...
+                (curr_procs_before
+                 - quantity)
+            if not self.log.check_log(r_msg, r_condition, False):
+                e_msg = 'Error removing %s dedicated cpus from partition %s.' \
+                        % (quantity, linux_machine.partition)
+                self.log.error(e_msg)
+                raise TestException(e_msg)
+
+        elif flag == "m":
+
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:253]: same here.

File: dlpar/dlpar_api/api.py [hunk 13]
```diff
@@ -451,3 +629,552 @@ def set_virtual_proc_and_proc_units(self, linux_machine, ideal_procs,
...
+    def __init__(self, config_payload, log='memory.log'):
+
+        TestCase.__init__(self, log, 'Memory', config_payload)
+
+        # Get test configuration
+        self.quant_to_test = 1024
+        self.quant_to_test = config_payload.get('mem_quantity_to_test')
+        self.sleep_time = 60
+        self.sleep_time = config_payload.get('sleep_time')
```
Comment (sacsant) [inline dlpar/dlpar_api/api.py:1014]: Values are overwritten by consecutive assignments. One of the assignment should be removed.

**Review by sacsant — CHANGES_REQUESTED** [review]

**Comment by SamirMulani** [comment]
Run Logs for DLPAR Dedicated and Shared mode,

**Dedicated mode Logs:**
```
[root@ltcden10-lp7 dlpar]# avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : a3728a04f70613a9359a34b3840c5faaceaf9d07
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-01T12.57-a3728a0/job.log
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-dedicated-c24d: STARTED
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-dedicated-c24d: PASS (268.87 s)
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-dedicated-c24d: STARTED
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-dedicated-c24d: PASS (192.31 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-01T12.57-a3728a0/results.html
JOB TIME   : 471.25 s
```

**Shared mode Logs**
```
 avocado run dlpar_main.py -m dlpar_main.py.data/dlpar.yaml --max-parallel-tasks=1
JOB ID     : a2573e2234017be05a1ea5aca15041148c31511b
JOB LOG    : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-07T00.27-a2573e2/job.log
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-shared-f54f: STARTED
 (1/2) dlpar_main.py:DlparTests.test_cpu_dlpar;run-config-lpar_mode-shared-f54f: PASS (384.48 s)
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-shared-f54f: STARTED
 (2/2) dlpar_main.py:DlparTests.test_mem_dlpar;run-config-lpar_mode-shared-f54f: PASS (193.38 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Workspace/AV/avocado-fvt-wrapper/results/job-2023-03-07T00.27-a2573e2/results.html
JOB TIME   : 580.21 s
```

File: dlpar/dlpar_main.py [hunk 14]
```diff
@@ -55,105 +62,78 @@ def run_cmd(self, test_cmd, dlpar_type_flag=""):
...
-                self.run_cmd(test_cmd, "smt:kill_process")
-            elif dlpar_type_flag == "cpu_fold":
-                self.run_cmd(test_cmd, "cpu_fold:kill_process")
+        for line in process.system_output('lparstat -i', ignore_status=True,
+                                          shell=True,
+                                          sudo=True).decode("utf-8") \
+                                                    .splitlines():
+            if component in line:
+                a = line.split(':')[-1].strip()
+                print(a)
```
[the remark this answers is not in the document]
Comment (PraveenPenguin) [reply dlpar/dlpar_main.py:88]: @SamirMulani please address this comment

File: dlpar/dlpar_main.py.data/dedicated.yaml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+cfg_cpu_per_proc: 8
+hmc_manageSystem:
+hmc_user:
+hmc_passwd:
+target_lpar_name:
+target_partition:
```
Comment (abdhaleegit) [inline dlpar/dlpar_main.py.data/dedicated.yaml:6]: rename to target_lpar_hostname  and lpar_name for hmc lpar name

File: dlpar/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply dlpar/README.md:26]: the statement is confusing

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
Please look into last 2 review comments

File: dlpar/dlpar_api/api.py (hunk 13 above)
Comment (SamirMulani) [reply dlpar/dlpar_api/api.py:1003]: The above changes are addressed in the latest commit

File: dlpar/dlpar_api/api.py (hunk 10 above)
Comment (SamirMulani) [reply dlpar/dlpar_api/api.py:645]: The above changes are addressed in the latest commit

File: dlpar/dlpar_api/api.py (hunk 9 above)
Comment (SamirMulani) [reply dlpar/dlpar_api/api.py]: This line is not needed, Removed in the latest commit.

File: dlpar/dlpar_api/api.py (hunk 8 above)
Comment (SamirMulani) [reply dlpar/dlpar_api/api.py]: This line is not needed, Removed in the latest commit.

File: dlpar/dlpar_api/api.py (hunk 5 above)
Comment (SamirMulani) [reply dlpar/dlpar_api/api.py]: Yes, Bothe variables are the same and are needed in multiple places.

File: dlpar/dlpar_main.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (SamirMulani) [reply dlpar/dlpar_main.py:88]: Removed in the latest commit.

File: dlpar/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamirMulani) [reply dlpar/README.md:26]: No Both are different. It is addressed in a new commit with a unique name.

**Comment by SamirMulani** [comment]
Please find the new pull request for the above DLPAR implementation.
https://github.com/avocado-framework-tests/avocado-misc-tests/pull/2578

**Pull request closed without merging**